src/rust/engine/fs/src/directory.rs

```diff
@@ -745,42 +879,88 @@ impl DigestTrie {
   }
 
   /// Return the Entry at the given relative path in the trie, or None if no such path was present.
+  /// If a directory component is a symlink, will follow the symlink. In cases where a symlink points
+  /// to a parent or current dir, the return may be None due to exceeding the link depth.
+  ///
+  /// Cannot follow a symlink above `self` (returns None).
   ///
-  /// An error will be returned if the given path attempts to traverse below a file entry.
+  /// An error will be returned if the given path attempts to traverse below a file entry or symlink.
   pub fn entry<'a>(&'a self, path: &Path) -> Result<Option<&'a Entry>, String> {
+    self.entry_helper(self, path, 0)
+  }
+
+  fn entry_helper<'a>(
+    &'a self,
+    root: &'a DigestTrie,
+    requested_path: &Path,
+    link_depth: LinkDepth,
+  ) -> Result<Option<&'a Entry>, String> {
     let mut tree = self;
     let mut path_so_far = PathBuf::new();
-    let mut components = path.components().peekable();
+    // Identical to path_so_far, but doesn't have components for "CurDir" symlinks
+    // E.g. If path_so_far is "dir/self/self/foo" and "dir/self -> .", then logical_path will be
+    // "dir/foo".
+    let mut logical_path = PathBuf::new();
+    let mut components = requested_path.components();
+    let mut current_entry: Option<&Entry> = None;
     while let Some(component) = components.next() {
+      if component == Component::CurDir {
+        // NB: This only happens if "." is the first component in a path.
+        continue;
+      } else if let Some(Entry::File(_)) = current_entry {
+        return Err(format!(
+          "{tree_digest:?} cannot contain a path at {requested_path:?}, \
+          because a file was encountered at {path_so_far:?}.",
+          tree_digest = self.compute_root_digest()
+        ));
+      } else if let Some(Entry::Directory(d)) = current_entry {
+        tree = &d.tree;
+      }
+
       path_so_far.push(component);
-      let component = component.as_os_str();
+      logical_path.push(component);
+      if component == Component::ParentDir {
```

Comment:
This very well could be a `while` loop, but the perf seems negligible and I'm tired of staring at the code.